api/src/test/java/org/openmrs/api/impl/DiagnosisServiceImplTest.java

```diff
@@ -0,0 +1,211 @@
+		}
+		if(encounterService == null){
+			encounterService = Context.getEncounterService();
+		}
+		executeDataSet(DIAGNOSIS_XML);
+	}
+
+	/**
+	 * @see DiagnosisService#save(Diagnosis) 
+	 */
+	@Test
+	public void saveDiagnosis_shouldSaveNewDiagnosis(){
+		String uuid = "a303bbfb-w5w4-25d1-9f11-4f33f99d456r";
+		Condition condition = conditionService.getConditionByUuid("2cc6880e-2c46-15e4-9038-a6c5e4d22fb7");
+		Encounter encounter = encounterService.getEncounterByUuid("y403fafb-e5e4-42d0-9d11-4f52e89d123r");
+		Patient patient = patientService.getPatient(2);
+		Diagnosis diagnosis = new Diagnosis();
+		diagnosis.setUuid(uuid);
+		diagnosis.setEncounter(encounter);
+		diagnosis.setCondition(condition);
+		diagnosis.setCertainty(ConditionVerificationStatus.CONFIRMED);
+		diagnosis.setPatient(patient);
+		diagnosis.setRank(2);
+		diagnosisService.save(diagnosis);
+		
+		Diagnosis savedDiagnosis = diagnosisService.getDiagnosisByUuid(uuid);
+		
+		Assert.assertEquals(uuid, savedDiagnosis.getUuid());
+		Assert.assertEquals(condition, savedDiagnosis.getCondition());
+		Assert.assertEquals(encounter, savedDiagnosis.getEncounter());
+		Assert.assertEquals(patient, savedDiagnosis.getPatient());
+		Assert.assertEquals(ConditionVerificationStatus.CONFIRMED, savedDiagnosis.getCertainty());
+		Assert.assertEquals(new Integer(2), savedDiagnosis.getRank());
+	}
+
+	/**
+	 * @see DiagnosisService#getDiagnosisByUuid(String) (String)
+	 */
+	@Test
+	public void getDiagnosisByUuid_shouldFindDiagnosisGivenValidUuid() {
+		String uuid="68802cce-6880-17e4-6880-a68804d22fb7";
+		Diagnosis diagnosis = diagnosisService.getDiagnosisByUuid(uuid);
+		Assert.assertEquals(uuid, diagnosis.getUuid());
+	}
+
+	/**
+	 * @see DiagnosisService#getDiagnosis(Integer)
+	 */
+	@Test
+	public void getDiagnosisById_shouldFindDiagnosifGivenId() {
+		Integer diagnosisId = 1;
+		Diagnosis diagnosis = diagnosisService.getDiagnosis(diagnosisId);
+		Assert.assertEquals("68802cce-6880-17e4-6880-a68804d22fb7", diagnosis.getUuid());
+		Assert.assertEquals(ConditionVerificationStatus.CONFIRMED, diagnosis.getCertainty());
+		Assert.assertEquals(diagnosisId, diagnosis.getDiagnosisId());
+	}
+
+	/**
+	 * @see DiagnosisService#getPrimaryDiagnoses(Encounter) 
+	 */
+	@Test
+	public void getPrimaryDiagnoses_shouldGetPrimaryDiagnoses(){
+		Encounter encounter = encounterService.getEncounterByUuid("y403fafb-e5e4-42d0-9d11-4f52e89d123r");
+		List<Diagnosis> diagnoses = diagnosisService.getPrimaryDiagnoses(encounter);
+		Assert.assertEquals(1, diagnoses.size());
+		Assert.assertEquals("68802cce-6880-17e4-6880-a68804d22fb7", diagnoses.get(0).getUuid());
+		Assert.assertEquals(ConditionVerificationStatus.CONFIRMED, diagnoses.get(0).getCertainty());
+		Assert.assertEquals(new Integer(1), diagnoses.get(0).getDiagnosisId());
+	}
+
+	/**
+	 * @see DiagnosisService#getDiagnoses(Patient, Date) 
+	 */
+	@Test
+	public void getDiagnoses_shouldGetDiagnosesOfPatient(){
```

Comment:
Did you add the test which takes care of the fromDate? Meaning that the results are different depending on the fromDate?

Reply:
I have modified it to capture different data.

Reply:
Where are the tests which expect different data depending on the fromDate parameter?

Reply:
Since the changes were done in the HibernateDiagnosisDAO, I added the tests in the HibernateDiagnosisDAOTest file

Reply:
The service layer also needs these tests. Just in case the implementation of the service layer ignored them (like you had originally done) instead of passing them down to the service layer. If you had such a test, then i would not have had to remind you about the un used fromDate parameter.

Reply:
Okay. I'm on it

Reply:
Done @dkayiwa